### PR TITLE
Improve monitoring ingest logic

### PR DESCRIPTION
**Pull request by lgarvey**
- add tests
- fail hard/stop ingesting data on exception
- track which files have already been processed

### CODE REVIEW

File: dnb_direct_plus/tests/test_management.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+        assert MonitoringFileRecord.objects.count() == 0
+
+    @pytest.mark.parametrize('archive_file',[
+        True,
+        False
+    ])
+    def test_files_are_archived(self, archive_file, mocker, settings):
+
+        settings.DNB_ARCHIVE_PROCESSED_FILES = archive_file
```
Comment (currycoder) [inline dnb_direct_plus/tests/test_management.py]: The `override_settings` decorator is normally a bit safer and will guarantee that settings changes are isolated to this test case: https://docs.djangoproject.com/en/3.0/topics/testing/tools/#django.test.override_settings

File: dnb_direct_plus/admin.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from django.contrib import admin
+
+from .models import MonitoringFileRecord
+
+
+@admin.register(MonitoringFileRecord)
+class UserAdmin(admin.ModelAdmin):
```
[the remark this answers is not in the document]
Comment (lgarvey) [reply dnb_direct_plus/admin.py]: Copy paste fail!

File: dnb_direct_plus/migrations/0001_initial_squashed_0002_auto_20191209_1402.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# Generated by Django 2.2.4 on 2019-12-09 14:03
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    replaces = [('dnb_direct_plus', '0001_initial'), ('dnb_direct_plus', '0002_auto_20191209_1402')]
```
[the remark this answers is not in the document]
Comment (lgarvey) [reply dnb_direct_plus/migrations/0001_initial_squashed_0002_auto_20191209_1402.py]: Yep, this makes sense.

File: dnb_direct_plus/tests/test_management.py (hunk 1 above)
Comment (lgarvey) [reply dnb_direct_plus/tests/test_management.py]: I read this comment and had a mild panic about the number of times I've used the pytest settings fixture  to modify settings. It looks like it's safe to do this: 

https://pytest-django.readthedocs.io/en/latest/helpers.html#settings

```This fixture will provide a handle on the Django settings module, and automatically revert any changes made to the settings (modifications, additions and deletions).```

File: dnb_direct_plus/management/commands/process_monitoring_data.py [hunk 4]
```diff
@@ -28,48 +28,51 @@ def _list_files(self):
...
-                    self._delete_file(file_name)
-
-        self.stdout.write('\n'.join(
-            '{file}\t\tTotal: {total}\tFailed: {failed}'.format(**line) for line in summary
-        ))
+
+            total, total_success = handler(bucket_path)
+
+            summary.append(
+                dict(file=file_name, total=total, failed=total - total_success)
```
[the remark this answers is not in the document]
Comment (lgarvey) [reply dnb_direct_plus/management/commands/process_monitoring_data.py]: Yes, there are still multiple conditions that will to an update not being applied, such as getting an update/diff without source data being stored against a company (should be an edge case), or if the timestamp indicates that the update is older than the last applied update.  But if something unexpected happens, it should stop processing at that point.

File: dnb_direct_plus/tests/test_management.py (hunk 1 above)
Comment (currycoder) [reply dnb_direct_plus/tests/test_management.py]: Oh my bad - sorry! I didn't spot that `settings` was a fixture parameter - I thought it was a standard import.

File: dnb_direct_plus/management/commands/process_monitoring_data.py [hunk 5]
```diff
@@ -28,48 +28,51 @@ def _list_files(self):
...
 
-            if not file_name.startswith(settings.DNB_MONITORING_REGISTRATION_REFERENCE):
+            if (not file_name.startswith(settings.DNB_MONITORING_REGISTRATION_REFERENCE) or
+                'HEADER' in file_name or
+                file_name.startswith(settings.DNB_ARCHIVE_PATH)):
                 continue
 
-            if 'HEADER' in file_name:
+            if MonitoringFileRecord.objects.filter(file_name=file_name).exists():
+                self.stdout.write(f'{file_name} already processed; skipping')
```
Comment (mcbhenwood) [inline dnb_direct_plus/management/commands/process_monitoring_data.py]: :+1: for appropriate use of a semicolon 😊

File: dnb_direct_plus/models.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+from django.db import models
+
+
+class MonitoringFileRecord(models.Model):
+    timestamp = models.DateTimeField(auto_now_add=True)
+    file_name = models.CharField(max_length=255)
```
Comment (mcbhenwood) [inline dnb_direct_plus/models.py]: Is it worth making `file_name` unique as a db constraint? I see the logic elsewhere to 'skip' a file already processed but it might be good to underpin this architectural decision with a db rule. (Optional).

Comment (lgarvey) [reply dnb_direct_plus/models.py]: Yep, this makes sense. I'll add it.

File: dnb_direct_plus/tests/test_management.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+        mocked = mocker.patch('dnb_direct_plus.management.commands.process_monitoring_data.Command._list_files')
+        mocked.return_value = [file_name, file_name2]
+        mocked_handler = mocker.patch(
+            'dnb_direct_plus.management.commands.process_monitoring_data.process_notification_file')
+        mocked_handler.side_effect = Exception('Something has gone wrong.')
+
+        out = StringIO()
+
+        with pytest.raises(Exception):
```
[the remark this answers is not in the document]
Comment (lgarvey) [reply dnb_direct_plus/tests/test_management.py]: Sentry has our back here. I was talking about wiring up a pingdom alert - but the way this is structured would make it a bit challenging.  An option might be to raise a p2 alert if no data has been received for a period, say 3 days.  

It would be easier if the app recorded an exception/error state for the file in MonitoringFileRecord, then we could alert on it.  This would complicate the logic a bit though.  

Happy to discuss this further

File: dnb_direct_plus/management/commands/process_monitoring_data.py [hunk 8]
```diff
@@ -28,48 +28,51 @@ def _list_files(self):
...
     def handle(self, *args, **options):
-
         files = self._list_files()
 
         summary = []
 
         for file_name in files:
 
-            if not file_name.startswith(settings.DNB_MONITORING_REGISTRATION_REFERENCE):
+            if (not file_name.startswith(settings.DNB_MONITORING_REGISTRATION_REFERENCE) or
```
[the remark this answers is not in the document]
Comment (lgarvey) [reply dnb_direct_plus/management/commands/process_monitoring_data.py]: Yep, this has been done.

File: dnb_direct_plus/tests/test_management.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (currycoder) [reply dnb_direct_plus/tests/test_management.py]: TBH, I think that having the ingestion stop immediately and popping the exception/trace in sentry is enough for now.  That should allow us to get a sense for the errors that will occur.  I do like the idea of raising an alert if we haven't seen any data for a few days though!
